controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -319,14 +385,26 @@ func (ew *EndpointsWatcher) getServicePublisher(id ServiceID) (sp *servicePublis
 /// servicePublisher ///
 ////////////////////////
 
-func (sp *servicePublisher) updateEndpoints(newEndpoints *corev1.Endpoints) {
+func (sp *servicePublisher) updateEndpoints(newResource interface{}) {
 	sp.Lock()
 	defer sp.Unlock()
-	sp.log.Debugf("Updating endpoints for %s", sp.id)
 
-	for _, port := range sp.ports {
-		port.updateEndpoints(newEndpoints)
+	switch res := newResource.(type) {
+	case *corev1.Endpoints:
+		sp.log.Debugf("Updating endpoints for %s", sp.id)
+		for _, port := range sp.ports {
+			port.updateEndpoints(newResource)
+		}
+	case *dv1beta1.EndpointSlice:
+		sp.log.Debugf("Updating endpoint slice for %s", sp.id)
+		for _, port := range sp.ports {
+			if !isTargetPortInSlice(port.targetPort, res.Ports) {
```

Comment:
Technically different EndpointSlice objects for a service could have different ports defined, but this seems strange to me and I'm not sure exactly what it means.  My assumption is that EndpointsSlices would be sliced up by address, meaning that each EndpointSlice for a service would have the same ports defined but a different set of addresses.

Does this match your understanding?

Reply:
Hm, that's interesting. Since I have started working on EndpointSlices, ports have been a major headache to deal with and there's quite a bit of confusion around it from my side.

The way I have been thinking about this has been based on the documentation on Endpoints in client-go. Specifically:
```
// EndpointSubset is a group of addresses with a common set of ports. The
// expanded set of endpoints is the Cartesian product of Addresses x Ports.
// For example, given:
//   {
//     Addresses: [{"ip": "10.10.1.1"}, {"ip": "10.10.2.2"}],
//     Ports:     [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
//   }
// The resulting set of endpoints can be viewed as:
//     a: [ 10.10.1.1:8675, 10.10.2.2:8675 ],
//     b: [ 10.10.1.1:309, 10.10.2.2:309 ]
```

They say an Endpoints object is the union of all subsets, where a subset is defined as a group of addresses with common ports. By this logic, I assumed an EndpointSlice is simply a subset -- a group of addresses (within a given limit, default 100 of them) that use the same ports. I have rarely encountered any endpoints with more than a subset, however, and replicating it has been a bit difficult, so I never saw my assumptions play out in minikube.

The only mention of subsets in EndpointSlices is:
```
// EndpointSlice represents a subset of the endpoints that implement a service.
// For a given service there may be multiple EndpointSlice objects, selected by
// labels, which must be joined to produce the full set of endpoints.
```

What do you think? Is there any easy way to test our assumptions or am I missing something very obvious?
Also, just to confirm: my assumption all along has been that EndpointSlices are sliced based on the common ports they share, so different slices would hold addresses that match that respective slice's port.

Reply:
After thinking about this some more, I think you're right about this.